File: packages/model/features-bigram.ts
import { features, hash, tokenize } from './features.js';

const utf8 = new TextEncoder();
export const BIGRAM_FEATURE_VERSION = 'gpu-search-features-bigram25-v1';

/** Ordered adjacent tokens share the word table; boundaries are length-prefixed. */
export function bigramFeatures(text: string) {
  const base = features(text);
  const tokens = tokenize(text).slice(0, 32).map(token => Array.from(token).slice(0, 64).join(''));
  const bigramIds: number[] = [];
  for (let i = 1; i < tokens.length; i++) {
    const left = utf8.encode(tokens[i - 1]), right = utf8.encode(tokens[i]);
    const bytes = new Uint8Array(10 + left.length + right.length);
    const view = new DataView(bytes.buffer);
    bytes.set([98, 58]);
    view.setUint32(2, left.length, true);
    bytes.set(left, 6);
    view.setUint32(6 + left.length, right.length, true);
    bytes.set(right, 10 + left.length);
    bigramIds.push(hash(bytes));
  }
  return { ...base, bigramIds, bigramCount: bigramIds.length };
}

File: training/features_bigram.py
"""Portable NumPy reference for the isolated ordered-bigram pooled16 experiment.

This is a new feature contract, not a change to gpu-search-features-v1.
"""
import numpy as np
from training.features import features as unigram_features, tokens, hash_bytes
VERSION='gpu-search-features-bigram25-v1'

def features(text):
    base=unigram_features(text)
    ts=[token[:64] for token in tokens(text)[:32]]
    bigrams=[]
    for left,right in zip(ts,ts[1:]):
        a,b=left.encode('utf-8'),right.encode('utf-8')
        blob=b'b:'+len(a).to_bytes(4,'little')+a+len(b).to_bytes(4,'little')+b
        bigrams.append(hash_bytes(blob))
    return {**base,'bigramIds':bigrams,'bigramCount':len(bigrams)}

def encode_numpy(params,texts,stages=False):
    pooled=[]
    for text in texts:
        f=features(text)
        def mean(name,ids):
            return np.mean(params[name][ids],axis=0,dtype=np.float32) if ids else np.zeros(params['word'].shape[1],np.float32)
        word=mean('word',f['wordIds'])
        if f['bigramIds']:
            word=np.float32(.75)*word+np.float32(.25)*mean('word',f['bigramIds'])
        pooled.append(np.float32(.5)*(word+mean('char',f['charIds'])))
    x=np.asarray(pooled,dtype=np.float32).reshape(-1,params['word'].shape[1])
    norms=np.linalg.norm(x,axis=-1,keepdims=True)
    vectors=x/np.maximum(norms,np.float32(1e-8))
    return {'pooled':x,'norms':norms,'vectors':vectors} if stages else vectors

def candidate_vectors(params,candidates):
    result=[]
    for candidate in candidates:
        vectors=encode_numpy(params,[candidate['label'],*candidate.get('aliases',[])])
        valid=vectors[np.linalg.norm(vectors,axis=1)>=1e-8]
        vector=valid.mean(axis=0) if len(valid) else np.zeros(params['word'].shape[1],np.float32)
        if candidate.get('context'):
            vector=vector+np.float32(.25)*encode_numpy(params,[candidate['context']])[0]
        norm=np.linalg.norm(vector)
        result.append(vector/max(norm,1e-8))
    return np.asarray(result,dtype=np.float32)
